src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -1101,37 +1145,219 @@ private void PopulateParamsFromRestRequest(Dictionary<string, object?> parameter
+            return isAuthorized;
+        }
+
+        /// <summary>
+        /// Performs authorization checks on entity level permissions and field level permissions for every entity and field
+        /// referenced in a GraphQL mutation for the given client role.
+        /// </summary>
+        /// <param name="context">Middleware context.</param>
+        /// <param name="clientRole">Client role header value extracted from the middleware context of the mutation</param>
+        /// <param name="topLevelEntityName">Top level entity name.</param>
+        /// <param name="operation">Mutation operation</param>
+        /// <param name="inputArgumentName">Name of the input argument (differs based on point/multiple mutation).</param>
+        /// <param name="parametersDictionary">Dictionary of key/value pairs for the argument name/value.</param>
+        /// <exception cref="DataApiBuilderException">Throws exception when an authorization check fails.</exception>
+        private void AuthorizeEntityAndFieldsForMutation(
+            IMiddlewareContext context,
+            string clientRole,
+            string topLevelEntityName,
+            EntityActionOperation operation,
+            string inputArgumentName,
+            IDictionary<string, object?> parametersDictionary
+        )
+        {
+            if (context.Selection.Field.Arguments.TryGetField(inputArgumentName, out IInputField? schemaForArgument))
+            {
+                // Dictionary to store all the entities and their corresponding exposed column names referenced in the mutation.
+                Dictionary<string, HashSet<string>> entityToExposedColumns = new();
+                if (parametersDictionary.TryGetValue(inputArgumentName, out object? parameters))
+                {
+                    // Get all the entity names and field names referenced in the mutation.
+                    PopulateMutationEntityAndFieldsToAuthorize(entityToExposedColumns, schemaForArgument, topLevelEntityName, context, parameters!);
+                }
+                else
+                {
+                    throw new DataApiBuilderException(
+                            message: $"{inputArgumentName} cannot be null for mutation:{context.Selection.Field.Name.Value}.",
+                            statusCode: HttpStatusCode.BadRequest,
+                            subStatusCode: DataApiBuilderException.SubStatusCodes.BadRequest
+                        );
+                }
+
+                // Perform authorization checks at field level.
+                foreach ((string entityNameInMutation, HashSet<string> exposedColumnsInEntity) in entityToExposedColumns)
+                {
+                    if (!AreFieldsAuthorizedForEntity(clientRole, entityNameInMutation, operation, exposedColumnsInEntity))
+                    {
+                        throw new DataApiBuilderException(
+                            message: $"Unauthorized due to one or more fields in this mutation.",
```

Comment:
nit: would it make sense to include `entityNameInMutation` here to help user triangulate which part of the mutation is at fault?
Current response looks like:
```json
{
  "errors": [
    {
      "message": "Unauthorized due to one or more fields in this mutation.",
      "extensions": {
        "message": "Unauthorized due to one or more fields in this mutation.",
        "stackTrace": "<stack_trace>",
        "code": "AuthorizationCheckFailed"
      }
    }
  ],
  "data": {
    "createbooks": null
  }
}
```

for the mutation executing as `Anonymous`
```graphql
mutation authz2Test {
  createbooks(items:[
    {title:"magicTreehouse3", publishers:{name: "RandomApartmentPubHouse"}}
    ])
  {
    items{
      id,
      title,
      publisher_id
    }
  }
}
```

Reply:
Will take this up as another issue: https://github.com/Azure/data-api-builder/issues/2123. Discussed over chat - fine to mark the comment as resolved for now.